srv/src/main/java/my/bookshop/handlers/AdminServiceAddressHandler.java

```diff
@@ -0,0 +1,189 @@
+	}
+
+	// Delegate ValueHelp requests to S/4 backend, fetching current user's addresses from there
+	@On(entity = cds.gen.adminservice.Addresses_.CDS_NAME)
+	public Result readAddresses(CdsReadEventContext context) {
+		String businessPartner = context.getUserInfo().getAttributeValues("businessPartner").stream().findFirst()
+			.orElseThrow(() -> new ServiceException(ErrorStatuses.FORBIDDEN, MessageKeys.BUPA_MISSING));
+
+		// forward the existing where condition
+		Select<?> select = Select.from(externalAddresses).where(a -> a.BusinessPartner().eq(businessPartner));
+		Optional<CqnPredicate> where = context.getCqn().where();
+		if(where.isPresent()) {
+			select.where(CQL.and(select.where().get(), where.get()));
+		}
+
+		// forward the selected columns (in Java, we cannot use our simplified projection Addresses yet)
+		List<CqnSelectListItem> columns = context.getCqn().items();
+		boolean hasStar = columns.stream().anyMatch(CqnSelectListItem::isStar);
+		if (hasStar || columns.isEmpty()) {
+			select.columns(getRelevantColumns());
+		} else {
+			List<String> relevantColumnsList = Arrays.asList(getRelevantColumns());
+			select.columns(columns.stream().filter(i -> relevantColumnsList.contains(i.asValue().value().asRef().firstSegment())).collect(Collectors.toList()));
+		}
+
+		// use Cloud SDK resilience capabilities..
+		ResilienceConfiguration config = ResilienceConfiguration.of(AdminServiceAddressHandler.class)
+			.timeLimiterConfiguration(TimeLimiterConfiguration.of(Duration.ofSeconds(10)));
+
+		return ResilienceDecorator.executeSupplier(() ->  {
+			// ..to access the S/4 system in a resilient way..
+			return bupa.run(select);
+		}, config, (t) -> {
+			// ..falling back to the already replicated addresses in our own database
+			Select<?> selectDb = Select.copy(context.getCqn()).where(select.where().get());
+			return db.run(selectDb);
+		});
+	}
+
+	// Replicate chosen addresses from S/4 when filling orders.
+	@Before
+	public void patchAddressId(DraftPatchEventContext context, Stream<Orders> orders) {
+		String businessPartner = context.getUserInfo().getAttributeValues("businessPartner").stream().findFirst()
+			.orElseThrow(() -> new ServiceException(ErrorStatuses.FORBIDDEN, MessageKeys.BUPA_MISSING));
+
+		orders.forEach(order -> {
+			// check if the address was updated
+			String addressId = order.getShippingAddressAddressID();
+			if(addressId != null) {
+				Result replica = db.run(Select.from(addresses).where(a -> a.BusinessPartner().eq(businessPartner).and(a.AddressID().eq(addressId))));
+				// check if the address was not yet replicated
+				if(replica.rowCount() < 1) {
+					Result remoteAddresses = bupa.run(Select.from(externalAddresses).columns(getRelevantColumns())
+							.where(a -> a.BusinessPartner().eq(businessPartner).and(a.AddressID().eq(addressId))));
+
+					if(remoteAddresses.rowCount() == 1) {
+						db.run(Upsert.into(addresses).entries(remoteAddresses));
```

Comment:
Do we want to advertise `Upsert` here? I thought, ideally we want to get rid of this.

Reply:
For these replication scenarios I think upsert is the ideal statement. It ensures that the replicated data is _**replaced**_ with the latest version from the remote system. So far we only had issues with `Upsert` on OData API level and with managed entities, but here in that case none of these two is the case.

Reply:
But here we explicitly check if the `replica` already exists. So in my opinion we could always use an `Insert` here. 

I'm also thinking about if it would make sense to do the insert/upsert in a separate transaction, because otherwise a parallel transaction could also insert the `remoteAddresses` and this transaction would fail when it is commited.

Reply:
You are right, in my previous comment I was thinking about the usage of Upsert in the `updateBusinessPartnerAddresses` method.

However from my perspective the `Upsert` statement expresses the idempotency that you are referring to. I'd like to avoid to have to start a separate transaction here and would hope that this "last one wins" is ensured by the `Upsert` statement.

Reply:
I don't think that the `Upsert` statement can guarantee this without opening a separate transaction, which is not the case I think. (Please correct me if I'm wrong @agoerler) If the `remoteAddress` does not exist when the transaction is opened, the conflict will only become visible when the transaction is committed.

Reply:
True, however with SQLite we are currently limited to a single connection, and can't open nested transactions here at the moment. We should revisit this, once we switch to H2 :) I will add a comment about this to the code.